### PR TITLE
[llm]support dpo pp

### PR DESCRIPTION
**Pull request by lugimzzz**
### PR types
 New features

### PR changes
APIs

### Description
1. 重构DPOTrainer与原版逐位对齐loss、metric
2. 重构DPOTrainer能够支持pp & vpp
3. 支持LoRA和多种DPO变体（包含KTO、DPO、ORPO、Simpo）
4. 新增支持多个开源模型

### CODE REVIEW

File: paddlenlp/transformers/sequence_parallel_utils.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import paddle
+from paddle import distributed as dist
+from paddle.autograd import PyLayer
+from paddle.distributed import fleet
+
+####################################################
```
Comment (ZHUI) [inline paddlenlp/transformers/sequence_parallel_utils.py:21]: 这个paddle里面有吗。我记得之前用paddle里面替换过一次。

Comment (lugimzzz) [reply paddlenlp/transformers/sequence_parallel_utils.py:21]: 暂时没有

File: paddlenlp/trl/dpo_criterion.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+    def __init__(self, config, use_infohub=False):
+        super(DPOCriterion, self).__init__()
+        self.config = config
+        if getattr(self.config, "dpo_config", None) is None:
+            raise ValueError("DPO Criterion requires model_config.dpo_config.")
+        self.dpo_config = copy.deepcopy(config.dpo_config)
+        if (
+            getattr(self.config, "tensor_parallel_output", False)
+            and getattr(self.config, "tensor_parallel_degree", 1) > 1
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py]: 这两个参数应该都会有了，config 内置了

File: paddlenlp/trl/dpo_criterion.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+            self.logprobs = ParallelCrossEntropy()
+        else:
+            self.logprobs = nn.CrossEntropyLoss(reduction="none")
+        self.use_infohub = use_infohub
+
+    def dpo_loss(self, policy_chosen_logps, policy_rejected_logps, reference_chosen_logps, reference_rejected_logps):
+        """DPO Loss"""
+        pi_logratios = policy_chosen_logps - policy_rejected_logps
+        ref_logratios = reference_chosen_logps - reference_rejected_logps
+        logits = pi_logratios - ref_logratios
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py:57]: `logits`这个变量，好像 kto 之后的方法就没有再使用了。对于一些方法，这些计算是多余的

File: paddlenlp/trl/dpo_criterion.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        rejected_labels,
+        response_indexs,
+        average_log_prob=False,
+    ):
+        """DPO logprobs"""
+        use_fused_head_and_loss_fn = getattr(self.config, "use_fused_head_and_loss_fn", False)
+        use_sparse_head_and_loss_fn = getattr(self.config, "use_sparse_head_and_loss_fn", False)
+        tensor_parallel_degree = getattr(self.config, "tensor_parallel_degree", 1)
+        tensor_parallel_output = getattr(self.config, "tensor_parallel_output", False)
+        sequence_parallel = getattr(self.config, "sequence_parallel", False)
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py]: 这三个变量dev，都内置了，一定有。

File: paddlenlp/trl/dpo_criterion.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+                hidden_states = paddle.take_along_axis(hidden_states, sparse_tgt_idx, axis=0)
+                hidden_states = AllGatherVarlenOp.apply(hidden_states)
+            else:
+                labels = labels.flatten()
+                sparse_tgt_idx = paddle.nonzero(labels != 0).flatten()
+                labels = paddle.take_along_axis(labels, sparse_tgt_idx, axis=0)
+
+                hidden_states = hidden_states.reshape([-1, hidden_states.shape[-1]])
+                hidden_states = paddle.take_along_axis(hidden_states, sparse_tgt_idx.unsqueeze(-1), axis=0)
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py:159]: 这里如果，只使用 tensor_parallel_degree 不使用 sequence_parallel 的话，会不会有问题？

Comment (ZHUI) [reply paddlenlp/trl/dpo_criterion.py:159]: labels hidden_states 都没切，应该没问题。

File: paddlenlp/trl/dpo_criterion.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+                transpose_y,
+                self.config.vocab_size,
+                tensor_parallel_degree,
+                tensor_parallel_output,
+                False,  # fused_linear
+                chunk_size,
+                return_token_loss=True,
+                ignore_index=0,
+            )
+        elif use_sparse_head_and_loss_fn:
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py:177]: 这个是不是可以默认开？

File: paddlenlp/trl/dpo_criterion.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+            per_token_logps = -self.logprobs(logits, labels)
+        else:
+            logits = logits.astype("float32")
+            if logits.shape[:-1] != labels.shape:
+                raise ValueError("Logits (batch and sequence length dim) and labels must have the same shape.")
+            # bs, seq
+            per_token_logps = -self.logprobs(logits, labels.unsqueeze(2)).squeeze(2)
+
+        if len(response_indexs.shape) == 3:
+            response_indexs = response_indexs[0]
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py:196]: 加个注释吧，`== 3` 略微hack。

File: paddlenlp/trl/dpo_criterion.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+                [(per_token_logps[response_index[1] : response_index[2]]).sum() for response_index in response_indexs],
+                axis=0,
+            )
+            rejected_logps = paddle.stack(
+                [(per_token_logps[response_index[2] : response_index[3]]).sum() for response_index in response_indexs],
+                axis=0,
+            )
+        else:
+            if get_env_device() == "npu":
+                chosen_list = []
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py:208]: npu 这个是为了性能优化吗？ 留个todo给npu同学，让他们后续统一吧。

File: paddlenlp/trl/dpo_criterion.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            average_log_prob = False
+        if reference_chosen_logps is None or reference_rejected_logps is None:
+            reference_chosen_logps, reference_rejected_logps, sft_loss = self.dpo_logps(
+                logits, chosen_labels, rejected_labels, response_indexs, average_log_prob
+            )
+            if self.use_infohub:
+                infohub.reference_chosen_logps.append(reference_chosen_logps)
+                infohub.reference_rejected_logps.append(reference_rejected_logps)
+                # pipeline mode requires return loss when self._compute_loss is True
+                return paddle.zeros([1])
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py:279]: emm, reference_model没有loss是不是?

File: paddlenlp/trl/dpo_criterion.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+        logits,
+        labels,
+    ):
+        """Forward"""
+        chosen_labels, rejected_labels, response_indexs, reference_chosen_logps, reference_rejected_logps = labels
+        if self.dpo_config.loss_type in ["ipo", "or", "simpo"]:
+            average_log_prob = True
+        else:
+            average_log_prob = False
+        if reference_chosen_logps is None or reference_rejected_logps is None:
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_criterion.py:271]: 这里为什么 `dpo_logps` `dpo_loss` 需要放到一个 forward中计算，有什么好处？

File: paddlenlp/trl/dpo_trainer.py [hunk 11]
```diff
@@ -38,230 +45,151 @@ class DPOTrainer(Trainer):
...
+            "dpop",
+            "or",
+            "simpo",
+        ]:
+            raise ValueError(f"Unknown loss type: {self.dpo_config.loss_type}")
+        if self.dpo_config.reference_free:
+            if ref_model is not None:
+                raise ValueError("reference_free set to True. No need to pass ref_model")
+            if self.dpo_config.loss_type not in ["sigmoid", "hinge", "ipo", "or", "simpo"]:
+                raise ValueError(f"{self.dpo_config.loss_type} does not support reference_free")
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_trainer.py:93]: 不使用lora的情况下，也可以不要 reference ？

File: paddlenlp/trl/dpo_trainer.py [hunk 12]
```diff
@@ -38,230 +45,151 @@ class DPOTrainer(Trainer):
...
+                reference_chosen_logps = paddle.zeros([1])
+                reference_rejected_logps = paddle.zeros([1])
             else:
-                with paddle.no_grad():
-                    reference_chosen_logps, reference_rejected_logps = self.ref_model(**dpo_inputs)
-                dpo_inputs["reference_chosen_logps"] = reference_chosen_logps
-                dpo_inputs["reference_rejected_logps"] = reference_rejected_logps
-            loss, policy_chosen_logps, policy_rejected_logps = model(**dpo_inputs)
+                if self.dpo_config.lora:
+                    with paddle.no_grad():
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_trainer.py:141]: 需要 autocast 吗？

File: paddlenlp/trl/dpo_trainer.py [hunk 13]
```diff
@@ -38,230 +45,151 @@ class DPOTrainer(Trainer):
...
+                        self.disable_lora(model)
+                        model.eval()
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_trainer.py:143]: assert ref_model is None?

File: paddlenlp/trl/dpo_trainer.py [hunk 14]
```diff
@@ -340,112 +262,124 @@ def log(self, logs, **kwargs):
...
+                            response_index[0] -= i * per_device_train_batch_size
+                            concatenated_inputs["response_indexs"][i].append(response_index)
+                    concatenated_inputs["response_indexs"][i] = paddle.stack(concatenated_inputs["response_indexs"][i])
+                    if model._layers.config.use_sparse_head_and_loss_fn:
+                        last_batch_response_length = concatenated_inputs["response_indexs"][i][0, 1]
+                        concatenated_inputs["response_indexs"][i][:, 1:] -= last_batch_response_length
+
+        concatenated_inputs["reference_chosen_logps"] = None
+        concatenated_inputs["reference_rejected_logps"] = None
+
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_trainer.py:309]: 建议这一大堆，封装成一个函数。

File: paddlenlp/trl/dpo_trainer.py [hunk 15]
```diff
@@ -340,112 +262,124 @@ def log(self, logs, **kwargs):
...
+        self._pp_data_buffer = []
+        inputs, labels = model._prepare_pipeline_inputs_func(concatenated_inputs)
+        self.reset_dpo_infohub()
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_trainer.py]: 这个最好，是使用完清理。现在是开始的时候清理。

File: paddlenlp/trl/dpo_trainer.py [hunk 16]
```diff
@@ -340,112 +262,124 @@ def log(self, logs, **kwargs):
...
+            train_eval="eval",
         )
-        policy_chosen_logps, policy_rejected_logps = policy_chosen_logps.detach(), policy_rejected_logps.detach()
-        if not self.reference_free:
-            chosen_rewards = self.args.dpo_beta * (policy_chosen_logps - reference_chosen_logps)
-            rejected_rewards = self.args.dpo_beta * (policy_rejected_logps - reference_rejected_logps)
-        else:
-            chosen_rewards = self.args.dpo_beta * (policy_chosen_logps)
-            rejected_rewards = self.args.dpo_beta * (policy_rejected_logps)
+        self.log_metric(**metric_inputs)
```
Comment (ZHUI) [inline paddlenlp/trl/dpo_trainer.py:350]: 比如这里 `self.reset_dpo_infohub()`

File: paddlenlp/trl/dpo_criterion.py (hunk 6 above)
Comment (lugimzzz) [reply paddlenlp/trl/dpo_criterion.py:177]: 开源模型还没有支持 use_sparse_head_and_loss_fn

File: paddlenlp/trl/dpo_criterion.py (hunk 9 above)
Comment (lugimzzz) [reply paddlenlp/trl/dpo_criterion.py:279]: 对注释写了为了适配pp

File: paddlenlp/trl/dpo_criterion.py (hunk 10 above)
Comment (lugimzzz) [reply paddlenlp/trl/dpo_criterion.py:271]: 主要为了pp

File: paddlenlp/trl/dpo_trainer.py (hunk 11 above)
Comment (lugimzzz) [reply paddlenlp/trl/dpo_trainer.py:93]: 可以，orpo、simpo算法就是 reference free的

File: paddlenlp/trl/dpo_trainer.py (hunk 12 above)
Comment (lugimzzz) [reply paddlenlp/trl/dpo_trainer.py:141]: 在def prediction_step写了autocast

File: llm/config/qwen/AdvertiseGen/w8a8_ptq_argument.json [hunk 17]
```diff
@@ -1,26 +0,0 @@
-{
```
Comment (ZHUI) [inline llm/config/qwen/AdvertiseGen/w8a8_ptq_argument.json]: 量化相关的全删了？原来是lora吗？

File: llm/run_finetune.py [hunk 18]
```diff
@@ -337,14 +337,6 @@ def neft_post_hook(module, input, output):
     else:
         trans_func = partial(get_convert_example(model), tokenizer=tokenizer, data_args=data_args)
 
-    if data_args.zero_padding:
```
Comment (ZHUI) [inline llm/run_finetune.py:340]: why?

File: paddlenlp/transformers/gemma/modeling.py [hunk 19]
```diff
@@ -1524,9 +1524,7 @@ def forward(
 
         # if labels is None，means we need full output, instead of tensor_parallel_output
         # tensor_parallel_output is togather with ParallelCrossEntropy
-        tensor_parallel_output = (
-            self.config.tensor_parallel_output and labels is not None and self.config.tensor_parallel_degree > 1
-        )
```
Comment (ZHUI) [inline paddlenlp/transformers/gemma/modeling.py:1563]: emm，这里原来是为了区分算不算 loss的。现在是影响到了dpo吗？

File: paddlenlp/transformers/yuan/configuration.py [hunk 20]
```diff
@@ -31,7 +31,7 @@ def __init__(
         hidden_act="silu",
         model_max_length=8192,
         initializer_range=0.02,
-        tensor_parallel_output=True,
```
Comment (ZHUI) [inline paddlenlp/transformers/yuan/configuration.py:34]: 这个需要修改默认值吗?不是的话，不需要写这个。

File: llm/config/qwen/AdvertiseGen/w8a8_ptq_argument.json (hunk 17 above)
Comment (lugimzzz) [reply llm/config/qwen/AdvertiseGen/w8a8_ptq_argument.json]: 这个只针对指定数据集的没有必要所以删掉了

File: llm/run_finetune.py (hunk 18 above)
Comment (lugimzzz) [reply llm/run_finetune.py:340]: 不对zero_padding限制模型了

File: paddlenlp/transformers/gemma/modeling.py (hunk 19 above)
Comment (lugimzzz) [reply paddlenlp/transformers/gemma/modeling.py:1563]: 对，这个用有没有label判断会导致dpo输出都是tensor_parallel_output 为False的情况，这块之前考虑是推理，我查看过推理代码会强制把tensor_parallel_output 设为False

File: paddlenlp/transformers/yuan/configuration.py (hunk 20 above)
Comment (lugimzzz) [reply paddlenlp/transformers/yuan/configuration.py:34]: 因为组网里面的实现方法默认都是False的情况，所以我觉得默认为False更好

File: llm/alignment/dpo/run_dpo.py [hunk 21]
```diff
@@ -20,18 +20,24 @@
...
-from dpo_argument import DPODataArgument, DPOModelArgument, DPOTrainingArguments
-
-from paddlenlp.datasets import ZeroPaddingMapDataset, load_dataset
-from paddlenlp.trainer import (
-    IntervalStrategy,
-    PdArgumentParser,
-    get_last_checkpoint,
-    set_seed,
+from dpo_argument import (
+    DPOConfig,
```
Comment (ZHUI) [inline llm/alignment/dpo/run_dpo.py:24]: `DPOConfig,DPOTrainingArguments` 这些看是否加到主repo？

**Review by ZHUI — APPROVED** [review]
LGTM
